### PR TITLE
[NEW] Experimental Game Center (externalComponents implementation)

### PR DESCRIPTION
**Pull request by d-gubert**
Implementation of the experimental Game Center, showcasing the new ExternalComponents from the Apps-Engine

For the next iteration:
* Create a new streamer event for externalComponent added/removed

### CODE REVIEW

File: app/apps/server/orchestrator.js [hunk 1]
```diff
@@ -72,6 +72,24 @@ class AppServerOrchestrator {
...
+		const apps = this._manager.get().map((prl) => {
+			const info = prl.getInfo();
+			info.status = prl.getStatus();
+			return info;
+		});
+		const activatedApps = apps.filter(
+			({ status }) => status === 'manually_enabled'
+		).map(({ id }) => id);
+		const { externalComponentManager: { providedComponents } } = this._manager;
+		Array.from(providedComponents.keys()).forEach((id) => {
```
Comment (d-gubert) [inline app/apps/server/orchestrator.js]: @lolimay this check is not needed - the Apps-Engine will only provide components when an app is properly enabled

Comment (lolimay) [reply app/apps/server/orchestrator.js]: ok, thank you!

File: app/apps/client/extrernalComponent.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+					break;
+				case 'getRoomInfo':
+					const { name: roomName, _id: roomId } = Session.get(`roomData${ Session.get('openedRoom') }`);
+					let { members } = await APIClient.get('v1/groups.members', { roomId });
+
+					members = members.map(({ _id, username, status }) => ({
+						userId: _id,
+						username,
+						avatarUrl: `${ baseUrl }${ getUserAvatarURL(username) }`,
+						status,
```
Comment (d-gubert) [inline app/apps/client/extrernalComponent.js]: @lolimay do you recall why is the status property sent over to the game?

Comment (lolimay) [reply app/apps/client/extrernalComponent.js]: I previously thought that the FRVR team should need this property to know the users' status. But there seems that they don't use it now. Should we remove this property?

File: app/apps/client/utils/index.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+function getRandomInt(max) {
+	return Math.floor(Math.random() * Math.floor(max));
+}
+
+/**
+ * Generate a random string with the specified length.
+ * @param length the length for the generated random string.
+ */
+export function randomString(length) {
```
Comment (d-gubert) [inline app/apps/client/utils/index.js]: @lolimay meteor provides the `Meteor.random()` function, it may be possible to replace these functions with it

**Comment by d-gubert** [comment]
@lolimay we should provide a setting to enable/disable the Game Center (disabled by default), and show a warning under it stating that it is an experimental feature :)

File: app/apps/client/utils/index.js (hunk 3 above)
Comment (lolimay) [reply app/apps/client/utils/index.js]: Resolved in #15781

File: app/apps/server/communication/rest.js [hunk 4]
```diff
@@ -260,6 +270,60 @@ export class AppsRestApi {
...
+		});
+
+		this.api.addRoute('externalComponentEvent', { authRequired: true }, {
+			post() {
+				if (this.bodyParams.event && this.bodyParams.externalComponent
+					&& ['IPostExternalComponentOpened', 'IPostExternalComponentClosed'].includes(this.bodyParams.event)
+				) {
+					let result;
+					try {
+						const { event } = this.bodyParams;
```
Comment (ggazzo) [inline app/apps/server/communication/rest.js]: const { event, externalComponent } = this.bodyParams;

File: app/apps/client/gameCenter/gameContainer.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	await APIClient.post('apps/externalComponentEvent', {
+		event: 'IPostExternalComponentOpened',
+		externalComponent,
+	});
+});
+
+Template.GameContainer.onDestroyed(async () => {
+	const externalComponent = await getExternalComponent();
+
+	await APIClient.post('apps/externalComponentEvent', {
```
Comment (ggazzo) [inline app/apps/client/gameCenter/gameContainer.js]: remove `await`

File: app/apps/client/gameCenter/tabBarButton.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import { TabBar } from '../../../ui-utils/client';
+import { APIClient } from '../../../utils';
+
+import './gameCenter.html';
+
+Meteor.startup(async function() {
+	const { externalComponents } = await APIClient.get('apps/externalComponents');
+
+	if (externalComponents.length > 0) {
```
[the remark this answers is not in the document]
Comment (ggazzo) [reply app/apps/client/gameCenter/tabBarButton.js]: should we prepare/create an event to new externalComponents?

File: app/apps/client/RealAppsEngineUIHost.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import { getUserAvatarURL } from '../../utils/lib/getUserAvatarURL';
+
+export class RealAppsEngineUIHost extends AppsEngineUIHost {
+	constructor() {
+		super();
+
+		this._baseURL = document.baseURI.slice(0, -1);
+	}
+
+	async getClientRoomInfo() {
```
Comment (ggazzo) [inline app/apps/client/RealAppsEngineUIHost.js]: async getClientRoomInfo(rid)

File: app/apps/client/RealAppsEngineUIHost.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+		const { name: slugifiedName, _id: id } = Session.get(`roomData${ Session.get('openedRoom') }`);
```
Comment (ggazzo) [inline app/apps/client/RealAppsEngineUIHost.js]: Rooms.findOne(...)

File: app/apps/client/RealAppsEngineUIHost.js (hunk 7 above)
Comment (d-gubert) [reply app/apps/client/RealAppsEngineUIHost.js]: We cannot apply this change for a couple of reasons:

* This class extends the `AppsEngineUIHost` class from the apps-engine, which defines the `getClientRoomInfo` and `getClientUserInfo` methods _without any parameters on the signature_

* The `externalComponents` use the `AppsEngineUIClient` to communicate with the Rocket.Chat window, and _it does not have access to either room id or user id to provide to this function call_

File: app/apps/client/gameCenter/tabBarButton.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (lolimay) [reply app/apps/client/gameCenter/tabBarButton.js]: Will implement this feature later

File: app/apps/server/communication/rest.js [hunk 9]
```diff
@@ -260,6 +270,57 @@ export class AppsRestApi {
 			},
 		});
 
+		this.api.addRoute('groupMembers', { authRequired: true }, {
```
Comment (sampaiodiego) [inline app/apps/server/communication/rest.js]: can you guys please explain why is this endpoint needed? [there are others](https://rocket.chat/docs/developer-guides/rest-api/groups/members/) endpoints that return this information already.

if this is really needed I'd ask to not duplicate the code at least.

Comment (lolimay) [reply app/apps/server/communication/rest.js]: Compared to the original one, this endpoint bypasses the API rate limitation. Is there any suggestion to refactor it?

Comment (sampaiodiego) [reply app/apps/server/communication/rest.js]: why is it being rate limited? any reason to call it so often? maybe the UI can cache the result.

Comment (d-gubert) [reply app/apps/server/communication/rest.js]: It's been quite a long time since we needed to introduce this endpoint, right? Maybe it is worth it to test the previous implementation again for us to check if there are any other strategies we could follow (as @sampaiodiego suggested)

Comment (d-gubert) [reply app/apps/server/communication/rest.js]: It's been quite a long time since we needed to introduce this endpoint, right? Maybe it is worth it to test the previous implementation again for us to check if there are any other strategies we could follow (as @sampaiodiego suggested)

Comment (lolimay) [reply app/apps/server/communication/rest.js]: OK, I will try to implement it with the existing API endpoint.

Comment (lolimay) [reply app/apps/server/communication/rest.js]: Use cached result when reach the rate limitation now, @sampaiodiego can you please review it https://github.com/RocketChat/Rocket.Chat/pull/15123/commits/8f922b65429ace3a4d4a2879fbf58ce9fce52161?

Comment (sampaiodiego) [reply app/apps/server/communication/rest.js]: nice.. but there a lot of conflicts 🙈

Comment (lolimay) [reply app/apps/server/communication/rest.js]: @d-gubert we need to create a new apps-engine beta version that contains both uiKit and externalComponent feature. The `1.12.0-beta.2575` only contains uiKit.

File: app/apps/client/gameCenter/gameContainer.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
+import { Template } from 'meteor/templating';
+
+import { modal } from '../../../ui-utils';
```
Comment (sampaiodiego) [inline app/apps/client/gameCenter/gameContainer.js]: always try to import at least to the `client` or `server` level, it makes the autocomplete happier =)
```suggestion
import { modal } from '../../../ui-utils/client';
```

File: app/apps/client/gameCenter/invitePlayers.js [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+	},
+});
+
+Template.InvitePlayers.events({
+	'submit #invite-players, click .js-confirm'(e, instance) {
+		const { data: { name } } = instance;
+		const users = instance.selectedUsers.get().map(({ username }) => username);
+		const privateGroupName = `${ name.replace(/\s/g, '-') }-${ Random.id(10) }`;
+
+		call('createPrivateGroup', privateGroupName, users).then((result) => {
```
Comment (sampaiodiego) [inline app/apps/client/gameCenter/invitePlayers.js]: please prefer using `await` =)

File: app/apps/client/gameCenter/tabBar.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import { Meteor } from 'meteor/meteor';
+import { Tracker } from 'meteor/tracker';
+
+import { APIClient } from '../../../utils';
+import { TabBar } from '../../../ui-utils/client';
+import { settings } from '../../../settings';
```
Comment (sampaiodiego) [inline app/apps/client/gameCenter/tabBar.js]:
```suggestion
import { APIClient } from '../../../utils/client';
import { TabBar } from '../../../ui-utils/client';
import { settings } from '../../../settings/client';
```

File: app/ui-master/public/icons/omnichannel.svg [hunk 13]
```diff
@@ -1,3 +1,3 @@
-<svg id="livechat" viewBox="0 0 20 20">
+<svg id="omnichannel" viewBox="0 0 20 20">
```
Comment (sampaiodiego) [inline app/ui-master/public/icons/omnichannel.svg]: is this change on purpose? doesn't seem related

File: app/apps/client/RealAppsEngineUIHost.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
+import { Meteor } from 'meteor/meteor';
+import { Session } from 'meteor/session';
+import { AppsEngineUIHost } from '@rocket.chat/apps-engine/client/AppsEngineUIHost';
+
+import { Rooms } from '../../models';
+import { APIClient } from '../../utils';
```
Comment (sampaiodiego) [inline app/apps/client/RealAppsEngineUIHost.js]:
```suggestion
import { Rooms } from '../../models/client';
import { APIClient } from '../../utils/client';
```

File: app/ui-master/public/icons/omnichannel.svg (hunk 13 above)
Comment (lolimay) [reply app/ui-master/public/icons/omnichannel.svg]: Yes, @renatobecker edited the `icons.svg` to change the icon's id from livechat to omnichannel in this [commit](https://github.com/RocketChat/Rocket.Chat/commit/17b1535a16c18b95b0836a1e271bb40940dbe7fd#diff-501ea7c311e0971aaa6decbe11d7df79R200) directly. We added a new icon called `game.svg` and run update icons scripts in this commit. To avoid overwriting his changes, so I changed the name and id of the livechat icon and run the update icons scripts again.

Comment (sampaiodiego) [reply app/ui-master/public/icons/omnichannel.svg]: ok then =) thanks

File: app/apps/client/RealAppsEngineUIHost.js [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	getUserAvatarUrl(username) {
+		let avatarUrl = getUserAvatarURL(username);
+
+		if (!avatarUrl.startsWith('http') && !avatarUrl.startsWith('data')) {
+			avatarUrl = `${ this._baseURL }${ avatarUrl }`;
+		}
+
+		return avatarUrl;
```
Comment (sampaiodiego) [inline app/apps/client/RealAppsEngineUIHost.js]: return first
```suggestion
		const avatarUrl = getUserAvatarURL(username);

		if (!avatarUrl.startsWith('http') && !avatarUrl.startsWith('data')) {
			return `${ this._baseURL }${ avatarUrl }`;
		}

		return avatarUrl;
```

File: app/apps/client/gameCenter/gameCenter.js [hunk 16]
```diff
@@ -0,0 +1,110 @@
+import toastr from 'toastr';
+import { Template } from 'meteor/templating';
+import { ReactiveVar } from 'meteor/reactive-var';
+
+import { modal } from '../../../ui-utils';
+import { APIClient, t } from '../../../utils';
```
Comment (sampaiodiego) [inline app/apps/client/gameCenter/gameCenter.js]:
```suggestion
import { modal } from '../../../ui-utils/client';
import { APIClient, t } from '../../../utils/client';
```
